diff --git a/core/modules/views/src/Plugin/views/PluginBase.php b/core/modules/views/src/Plugin/views/PluginBase.php
index e478e9130..34c326327 100644
--- a/core/modules/views/src/Plugin/views/PluginBase.php
+++ b/core/modules/views/src/Plugin/views/PluginBase.php
@@ -360,11 +360,9 @@ protected function viewsTokenReplace($text, $tokens) {
       // No need to run filterAdmin on an empty string.
       return '';
     }
-    if (empty($tokens)) {
-      return Xss::filterAdmin($text);
-    }

     $twig_tokens = [];
+    if (!empty($tokens)) {
     foreach ($tokens as $token => $replacement) {
       // Twig wants a token replacement array stripped of curly-brackets.
       // Some Views tokens come with curly-braces, others do not.
@@ -400,11 +398,7 @@ protected function viewsTokenReplace($text, $tokens) {
         $twig_tokens[$top] += $token_array;
       }
     }
-
-    if ($twig_tokens) {
-      // Use the unfiltered text for the Twig template, then filter the output.
-      // Otherwise, Xss::filterAdmin could remove valid Twig syntax before the
-      // template is parsed.
+    }

       $build = [
         '#type' => 'inline_template',
@@ -422,10 +416,6 @@ function ($children, $elements) {
       // using Renderer::executeInRenderContext() instead.
       // @todo: https://www.drupal.org/node/2566621
       return (string) $this->getRenderer()->renderInIsolation($build);
-    }
-    else {
-      return Xss::filterAdmin($text);
-    }
   }

   /**
